h2o-r/h2o-package/R/zzz.R

```diff
@@ -1 +1,61 @@
+
+
+# Helper functions for h2o.explain commands in .onLoad
+# Inspired by vctrs' s3_register function for registering s3 methods for generics from suggested packages
+.s3_register <- function(package, generic, class) {
+  method_env <- if (isNamespace(topenv())) asNamespace(environmentName(topenv())) else parent.frame()
+  method <- get(paste(generic, class, sep = "."), envir = method_env)
+  
+  # Register hook in case package is unloaded & reloaded
+  setHook(packageEvent(package, "onLoad"),
+          function(...) {
+            registerS3method(generic, class, method, envir = asNamespace(package))
+          }
+  )
+  
+  # Don't register if the package is not present
+  if (!isNamespaceLoaded(package)) {
+    return(invisible())
+  }
+  
+  # Register iff generic exists in the package environment
+  if (exists(generic, asNamespace(package))) {
+    registerS3method(generic, class, method, envir = asNamespace(package))
+  }
+  
+  invisible()
+}
+
+
+# This was moved from connection.R; .onAttach, .onDetach, .Last are still in connection.R
+.onLoad <- function(lib, pkg) {
+  
+  # for file.download, used to download h2o.jar
+  options(timeout = max(1000, getOption("timeout")))
```

Comment:
I'd suggest to use the same pattern as is often used for `par`:

```r
old_options <- options(timeout = max(1000, getOption("timeout")))  # NOTE: old_options is a named list containing only the old values of the changed options
on.exit(options(old_options))
```

This way the timeout is set only within this function scope so we'd need to move the `.h2o.downloadJar()` inside this function. 

I think it shouldn't cause `.h2o.downloadJar()` to be ran more times than it used to but I am not sure; my assumptions are:
>Now it is called on every `source` which would be just once per session and this would change to be called just once per load which should also be just once per session (both statements should be valid unless a user decides to unload the pkg and reload it which should be relatively rare).

Reply:
I think it is better to move setting option to `.h2o.downloadJar`

Reply:
@tomasfryda Are you saying that you want to change things around so that the timeout only applies to the `.h2o.downloadJar()` function?  

@jangorecki Why do you suggest that we move the option setting to `.h2o.downloadJar`?  The same reason above (so it only applies to one function)?

Reply:
Here's the current behavior:

```
> library(h2o)

----------------------------------------------------------------------

Your next step is to start H2O:
    > h2o.init()

For H2O package documentation, ask for help:
    > ??h2o

After starting H2O, you can use the Web UI at http://localhost:54321
For more information visit https://docs.h2o.ai

----------------------------------------------------------------------


Attaching package: ‘h2o’

The following objects are masked from ‘package:stats’:

    cor, sd, var

The following objects are masked from ‘package:base’:

    &&, %*%, %in%, ||, apply, as.factor, as.numeric, colnames,
    colnames<-, ifelse, is.character, is.factor, is.numeric, log,
    log10, log1p, log2, round, signif, trunc

> getOption("timeout")
[1] 1000
> detach("package:h2o", unload=TRUE)
[1] "A shutdown has been triggered. "
> getOption("timeout")
[1] 1000
```
It will go back to 60 seconds only with a new R session.

Reply:
Yes, so current behavior can be improved

Reply:
@ledell yes, I would limit side effects like this to the smallest possible extent so we don't affect anything else, e.g., some other library the same session.